accepted/0000-workspaces-link-phase-3.md

```diff
@@ -0,0 +1,164 @@
+  "version": "20.0.3",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/facebook/jest.git"
+  },
+  "license": "BSD-3-Clause",
+  "main": "build/index.js",
+  "browser": "build-es5/index.js",
+  "dependencies": {
+    "chalk": "^1.1.3",
+    "pretty-format": "^20.0.3"
+  }
+}
+```
+
+jest-diff (workspace that refers jest-matcher-utils)
+```
+{
+  "name": "jest-diff",
+  "version": "20.0.3",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/facebook/jest.git"
+  },
+  "license": "BSD-3-Clause",
+  "main": "build/index.js",
+  "browser": "build-es5/index.js",
+  "dependencies": {
+    "chalk": "^1.1.3",
+    "diff": "^3.2.0",
+    "**jest-matcher-utils**": "^20.0.3",
+    "pretty-format": "^20.0.3"
+  }
+}
+```
+
+When user runs yarn install, this folder structure of the Workspace gets created
+```
+| jest/
+| ---- node_modules/
+| -------- chalk/
+| -------- diff/
+| -------- pretty-format/
+| ---- package.json
+| ---- packages/
+| -------- jest-matcher-utils/
+| ------------ node_modules/ (empty, all dependencies hoisted to the root)
+| ------------ package.json
+| -------- jest-diff/
+| ------------ node_modules/
+| ---------------- **jest-matcher-utils**/  (symlink) -> ../jest-matcher-utils
+| ------------ package.json
+...
+```
+
+### Dependencies and version matching
+
+Yarn would only link workspaces to each other if they match semver conditions.
+For example,
+
+* `jest-matcher-utils` package.json is `20.0.3`
+* if `jest-diff` package.json dependencies has `jest-matcher-utils` with version specifier that matches 20.0.3, e.g. `"^20.0.3"` then Yarn will make a link from `jest-diff/node_modules/jest-matcher-utils` to `jest-matcher-utils` workspace
+* if `jest-diff` package.json dependencies has `jest-matcher-utils` with version specifier that does not match `20.0.3`, e.g. `"^19.0.0"` then Yarn would fetch `jest-matcher-utils@^19.0.0` from npm registry and install it the regular way
+
+
+### Problems with peer dependencies and hoisting
+
+There is a common [peer dependency problem](http://codetunnel.io/you-can-finally-npm-link-packages-that-contain-peer-dependencies/) when using **yarn link** on local packages that people can work around in Node 6+ by setting **--preserve-symlinks** runtime flag.
+In Workspaces this situation won't be a problem because node_modules are installed in Workspace root and Node.js `require()` statements will resolve third-party peer dependencies by going up the folder tree and reaching the Workspaces' root node_modules.
+
+As long as **jest-matcher-utils** does not make relative requires via its parent folder, flag **--preserve-symlinks** won't be necessary.
+
+
+## Drawbacks
+
+This solution creates a symlink inside node_modules of a Workspace package and symlinks have multiple drawbacks:
+
+* Symlinks are not supported in all tools (e.g. watchman)
+* Symlinks are not supported well in all OS and environments (Windows pre 10 Creative updated, Docker on SMB storage(?))
```

Comment:
Could we have an alternative mode, especially for CI, where we run "yarn build" and then "yarn pack" on every package and cp the build into node_modules instead of using symlinks? CI's don't need the symlinks because you aren't changing the code there. Symlinks are only really important during development.